File: Python/Samples/Template/ProductFixingTemplate.py
from abc import ABC, abstractmethod

class ProductFixingTemplate(ABC):
    @abstractmethod
    def findWorkOptionLeg(self):
        """找出Working Option Leg"""
        pass

    @abstractmethod
    def checkBarriers(self):
        """檢查是否已經觸及生效/失效障礙"""
        pass
    
    @abstractmethod
    def rebateBarriers(self):
        """Return 是否繼續比價
        由檢核障礙結果做Rebate process
        """
        pass

    @abstractmethod
    def fixingOptionLeg(self):
        """計算每個部位的PayOff及交割結果"""
        pass

    @abstractmethod
    def checkTriggers(self):
        """檢核Triggers"""
        pass

    def fixing(self):
        """開始比價"""
        #Find the work option leg
        self.findWorkOptionLeg()
            
        #Check Barriers
        self.checkBarriers()

        #Rebate Barriers
        if(self.rebateBarriers()):
            self.fixingOptionLeg()
            self.checkTriggers()
        else:
            self.checkTriggers()

File: CSharp/DP.Domain/Samples/Template/ProductFixingTemplate.cs
namespace DP.Domain.Samples.Template
{
    public abstract class ProductFixingTemplate
    {
        /// <summary>
        /// 找出Working Option Leg
        /// </summary>
        protected abstract void FindWorkOptionLeg();
        /// <summary>
        /// 檢查是否已經觸及生效/失效障礙
        /// </summary>
        protected abstract void CheckBarriers();
        /// <summary>
        /// 由檢核障礙結果做Rebate process
        /// </summary>
        /// <returns>是否繼續比價</returns>
        protected abstract bool RebateBarriers();
        /// <summary>
        /// 計算每個部位的PayOff及交割結果
        /// </summary>
        protected abstract void FixingOptionLeg();
        /// <summary>
        /// 檢核Triggers
        /// </summary>
        protected abstract void CheckTriggers();

        /// <summary>
        /// 開始比價
        /// </summary>
        public void Fixing()
        {
            //Find the work option leg
            this.FindWorkOptionLeg();
            
            //Check Barriers
            this.CheckBarriers();

            //Rebate Barriers
            if(this.RebateBarriers())
            {
                //Fixing working Option Leg
                this.FixingOptionLeg();

                //Check Triggers
                this.CheckTriggers();
            }
            else
            {
                //Check Triggers
                this.CheckTriggers();
            }
        }
    }
}